repo/blob/sftp/sftp_storage_test.go

```diff
@@ -4,23 +4,130 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	"net"
 	"os"
+	"os/exec"
+	"path/filepath"
 	"strconv"
+	"strings"
 	"testing"
+	"time"
 
 	"github.com/kopia/kopia/internal/blobtesting"
 	"github.com/kopia/kopia/internal/testlogging"
+	"github.com/kopia/kopia/internal/testutil"
 	"github.com/kopia/kopia/repo/blob"
 	"github.com/kopia/kopia/repo/blob/sftp"
 )
 
+const (
+	dockerImage  = "atmoz/sftp"
+	dialTimeout  = 10 * time.Second
+	sftpUsername = "foo"
+)
+
+func mustGetLocalTmpDir(t *testing.T) string {
+	t.Helper()
+
+	tmpDir, err := ioutil.TempDir(".", ".creds")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tmpDir, err = filepath.Abs(tmpDir)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	t.Cleanup(func() {
+		os.RemoveAll(tmpDir)
+	})
+
+	return tmpDir
+}
+
+func mustRunOrSkip(t *testing.T, cmd string, args ...string) []byte {
+	t.Helper()
+
+	o, err := exec.Command(cmd, args...).CombinedOutput()
+	if err != nil {
+		t.Fatal(string(o), err)
+	}
+
+	t.Logf("output: %v", string(o))
+
+	return o
+}
+
+func startDockerSFTPServerOrSkip(t *testing.T, idRSA string) (host string, port int, knownHostsFile string) {
+	t.Helper()
+
+	tmpDir := mustGetLocalTmpDir(t)
+	sshHostED25519Key := filepath.Join(tmpDir, "ssh_host_ed25519_key")
+	sshHostRSAKey := filepath.Join(tmpDir, "ssh_host_rsa_key")
+
+	mustRunOrSkip(t, "ssh-keygen", "-t", "ed25519", "-P", "", "-f", sshHostED25519Key)
+	mustRunOrSkip(t, "ssh-keygen", "-t", "rsa", "-P", "", "-f", sshHostRSAKey)
+
+	// see https://github.com/atmoz/sftp for instructions
+	shortContainerID := testutil.RunContainerAndKillOnCloseOrSkip(t,
+		"run", "--rm", "-p", "0:22",
+		"-v", idRSA+".pub:/home/"+sftpUsername+"/.ssh/keys/id_rsa.pub:ro",
+		"-v", sshHostED25519Key+":/etc/ssh/ssh_host_ed25519_key:ro",
+		"-v", sshHostRSAKey+":/etc/ssh/ssh_host_rsa_key:ro",
+		"-d", dockerImage,
+		sftpUsername+"::::upload")
+	sftpEndpoint := testutil.GetContainerMappedPortAddress(t, shortContainerID, "22")
+
+	// wait for SFTP server to come up.
+	deadline := time.Now().Add(dialTimeout)
+	for time.Now().Before(deadline) {
+		t.Logf("waiting for SFTP server to come up on '%v'...", sftpEndpoint)
+
+		conn, err := net.DialTimeout("tcp", sftpEndpoint, time.Second)
+		if err != nil {
+			t.Logf("err: %v", err)
+			time.Sleep(time.Second)
+
+			continue
+		}
+
+		defer conn.Close()
+
+		parts := strings.Split(sftpEndpoint, ":")
+		host = parts[0]
+		port, _ = strconv.Atoi(parts[1])
+		knownHostsFile = filepath.Join(t.TempDir(), "known_hosts")
+		knownHostsData := mustRunOrSkip(t, "ssh-keyscan", "-t", "rsa", "-p", strconv.Itoa(port), host)
+
+		ioutil.WriteFile(knownHostsFile, knownHostsData, 0600)
+
+		t.Logf("SFTP server OK on host:%q port:%v. Known hosts file: %v", host, port, knownHostsFile)
+
+		return
+	}
+
+	t.Skipf("SFTP server did not start!")
+
+	return //nolint:nakedret
```

Comment:
What is the effect of removing this return?

Reply:
it does not compile.

Reply:
Oh, I see why.